### ISSUE TITLE
Generalizable instances

### ISSUE THREAD

**Issue by jespercockx** [agda/agda#3126]
This is perhaps a bit of a crazy request but I think having something like this would be very nice:
```agda
generalise
  {G} : Set
  instance
    isGroup : IsGroup G

my-id : G  -- type is generalized to `my-id : {G : Set} {{isGroup : IsGroup G}} -> G`
my-id = IsGroup.id
```

**Comment by UlfNorell** [comment]
cf #1151

**Comment by UlfNorell** [comment]
Also related #3275

**Comment by ajrouvoet** [comment]
A problem with generalize is that it reserves some (usually short) names. Another problem is that there is little control over where the generalization happens.

Perhaps a less radical, perhaps complementary solution to the problem of these type signatures that get out of control in the presence of type-class arguments, could be inspired by Scala's syntax for [context bounds](https://docs.scala-lang.org/tutorials/FAQ/context-bounds.html).

Where instead you would write something like:

```
my-id : (G <: IsGroup) -> G
```

Which desugars to `my-id : forall {G} {{_ : IsGroup G}} -> G`

**Comment by gallais** [comment]
Note that you can write this kind of things with the support for patterns in
telescopes:

```agda
record IsPointed (A : Set) : Set where
  field point : A

record Pointed : Set₁ where
  constructor !_
  field set           : Set
        {{isPointed}} : IsPointed set
open Pointed

f : ((! G) : Pointed) → G
f = {!!}
```

**Comment by ajrouvoet** [comment]
That is a neat trick. I'll have to play with that some more.

It is really tricky to work with bundled typeclasses *and* make instance search useful. The above requires the definition of a bundle that makes the unbundled class available in the instance scope. That means that we cannot also have something in the instance scope that builds the bundled structure from the unbundled stuff. That would mean that I cannot make the Pointed parameter an instance argument and have to construct the thing by myself, which misses the point a little bit.

I'm really interested in finding out how to get instance search to work for me in Agda for dealing with algebraic structures and have been struggling to get something that works consistently for a while. If others are interested in this as well, lets talk :)

**Comment by andreasabel** [comment]
Concerning the OP: currently we can write:
```agda
record IsGroup (G : Set) : Set where
  field
    id : G
open IsGroup {{...}}

variable
  G : Set

my-id : {{IsGroup G}} → G
my-id = id
```
I am tempted to close this issue.
It seems to be an instance of a more general feature that is for instance present in Rocq, that you can implicitly convert an algebraic structure like `Group` to its carrier.  (In @gallais spelling of the example, this would mean that `set` is an implicit coercion.)

**Comment by jespercockx** [comment]
I agree that implicit coercions would also solve this issue, but my (perhaps naive) hope with this feature request was that it would be perhaps an easier way to get a similar effect in this case. But I don't have a strong argument for it, so feel free to close this.
